### PR TITLE
Flexline width exceeds container width

### PR DESCRIPTION
**Pull request by AlexBalo**
In this PR I am trying to address an issue that occurs when the calculation of flexlines exceeds the permitted size of their container and some items can be shrunk.

After successfully calculating the number of flexline the FlexboxLayout proceeds its measurement by determining the main size through the determineMainSize() method. In certain situations, the width of some flexlines might be bigger than the size of their container. In this cases, we need to restrict their calculated width by retriggering a second layout calculation and enforcing their width as the one of their containers.

**Before the change**
The example below shows how the TextView width is calculated based on the full screen width causing its text to be cut out. In that specific case the total width of the flexline is: ImageView width (which has fixed size) + the TextView width (which is equals to the screen size). No re-measurement happens after the calculation of the number of flexlines since the main size calculated in the determineMainSize() is based on the widest flexline, which in this case exceeds the container width:

Screenshot 2019-12-04 at 10 49 58

**After the change**
After changing the determineMainSize() method to disallow lines to be wider than their container, the flexlines are correctly recalculated and laid out. The pictures below show the same example with both a single ImageView or with two ImageView on the left and right of the TextView. The end results are showing a correctly recalculated layout:

Screenshot 2019-12-04 at 10 49 10

Screenshot 2019-12-04 at 10 50 35

[This Gist](https://gist.github.com/AlexBalo/38684b94a9919c0106483e35df304546) contains the classes that have been used to reproduce the scenarios described above.

### CODE REVIEW

File: demo-cat-gallery/src/main/java/com/google/android/flexbox/apps/catgallery/perf/TextViewMeasurementBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          "yqy8GG/lWcnmlGcl4ZRnJdlV9pIGM61SPpaUySf1XBgQU+UofKBZAAAAAElFTkSuQmCC"
+
+  fun build(container: ViewGroup) {
+    val wrapper = defaultContainer()
+    val root = defaultContainer(FlexDirection.ROW, true)
+    val textView = textview()
+
+    root.addView(imageview())
+    root.addView(textView)
+//    root.addView(imageview())
```
Comment (AlexBalo) [inline demo-cat-gallery/src/main/java/com/google/android/flexbox/apps/catgallery/perf/TextViewMeasurementBuilder.kt]: It is possible to uncomment this line to verify the behavior with ImageView at the beginning and end of the TextView.

**Comment by AlexBalo** [comment]
@googlebot I signed it!

**Comment by AlexBalo** [comment]
I updated [the Gist](https://gist.github.com/AlexBalo/38684b94a9919c0106483e35df304546) to also include a Checkbox widget, because I noticed a different behavior happening in this case. In fact when Checkbox is allowed to shrink (flexShrink = 1f), during the recalculation can end up being shrunk more than what I would consider its "minimum size". 

In the image below it is possible to notice the behavior explained above:
 
Screenshot 2019-12-06 at 18 39 30

I am trying to understand if there is a viable way to take scenarios like this one into account during measurement.

**Comment by thagikura** [comment]
Thanks for the PR! Sorry for the delayed reply... 
The fix looks good to me! Can you please add a test to FlexboxHelperTest to verify the fix.

Looks like the issue for the ButtonDrawable can be fixed in another PR.
The code you suggested looks too specific for FlexboxLayout if it's part of a CustomView. So I'd look into if a View's minimum width/height can be used for determing the size to be shrunk.

**Comment by AlexBalo** [comment]
Hi @thagikura, thank you for reviewing the diff. 

I added a couple of tests to verify the fix. I agree we can address the ButtonDrawable in a separate PR.

**Comment by Ribesg** [comment]
This "fix" broke my UI. Please don't publish breaking changes as fix versions in the future.

**Comment by AlexBalo** [comment]
> This "fix" broke my UI. Please don't publish breaking changes as fix versions in the future.

Hi @Ribesg, sorry to hear this fix broke your UI. I had a problem with text rendering and based on the investigation in the Gist in the description, it looked like during onMeasure the Text widget were given the full-screen width.

May I ask you if you can share more information about your scenario that got broken after this fix? 

Thank you in advance.

**Pull request merged**